ocaml/tests/test_clustering.ml

```diff
@@ -376,12 +376,205 @@ let test_disallow_unplug_ro_with_clustering_enabled =
+  begin match Db.SR.get_PBDs ~__context ~self:sm_sr with
+  | [ pBD ] ->
+    Db.PBD.set_currently_attached ~__context ~self:pBD ~value:false;
+    Db.SR.destroy ~__context ~self:sm_sr
+  | _ -> Alcotest.fail "only one PBD should be plugged into this SR"
+  end;
+  choose_cluster_stack_should_select default_smapiv3 ~__context;
+
+  (* default_smapiv3 is the only common cluster stack *)
+  let _  : API.ref_SM = T.make_sm ~__context ~_type:"hba" ~required_cluster_stack:[ test_default ; default_smapiv3 ] () in
+  let _ = create_and_plug_sr ~_type:"hba" ~__context () in
+  choose_cluster_stack_should_select default_smapiv3 ~__context;
+
+  let _  : API.ref_SM = T.make_sm ~__context ~_type:"gfs2" ~required_cluster_stack:[ default_smapiv3 ; default_smapiv3 ] () in
+  let _ = create_and_plug_sr ~_type:"gfs2" ~__context () in
+  choose_cluster_stack_should_select default_smapiv3 ~__context;
+
+  let _ = create_and_plug_sr ~_type:"type_not_in_sm_table" ~__context () in
+  Alcotest.check_raises
+    "choose_cluster_stack should fail when checking SR with no matching SM type"
+    (Failure "SR type not found in SM table.")
+    (fun () -> Cluster_stack_constraints.choose_cluster_stack ~__context |> ignore)
+
+let test_choose_cluster_stack_with_sms_and_clusters () =
+  let __context = T.make_test_database () in
+  let _ = T.make_cluster_and_cluster_host ~__context ~cluster_stack:default_smapiv3 () in
+  let _  : API.ref_SM = T.make_sm ~__context ~_type:"lvm" ~required_cluster_stack:[ test_default ] () in
+  let _ = create_and_plug_sr ~_type:"lvm" ~__context () in
+  choose_cluster_stack_should_select test_default ~__context
+
+let test_choose_cluster_stack =
+  [ "test_choose_cluster_stack_clusters_no_sms", `Quick, test_choose_cluster_stack_clusters_no_sms
+  ; "test_choose_cluster_stack_with_sms_and_clusters", `Quick, test_choose_cluster_stack_with_sms_and_clusters
+  ; "test_choose_cluster_stack_sms_no_clusters", `Quick, test_choose_cluster_stack_sms_no_clusters
+  ]
+
+let get_ha_cluster_stack ~__context =
+  Db.Pool.get_ha_cluster_stack ~__context ~self:(Helpers.get_pool ~__context)
+
+let assert_cluster_stack_is cluster_stack ~__context =
+  Alcotest.(check string)
+    "Pool's ha_cluster_stacks"
+    cluster_stack
+    (get_ha_cluster_stack ~__context)
+
+let test_pool_ha_cluster_stacks_no_ha_no_clustering () =
+  let __context = T.make_test_database () in
+  (* HA disabled by default *)
+  assert_cluster_stack_is default ~__context
+
+let get_only_cluster_host ~__context : API.ref_Cluster_host =
+  match Db.Cluster_host.get_all ~__context with
+  | [ cluster_host ] -> cluster_host
+  | lst ->
+    let nodes = lst |> List.length |> string_of_int in
+    raise Api_errors.(Server_error (cluster_does_not_have_one_node, [ nodes ]))
+
+let test_pool_ha_cluster_stacks_no_ha_with_clustering () =
+  (* Test that cluster creation and destruction don't set
+   * HA cluster stacks, as HA is disabled *)
```

Comment:
They should in fact set it, because if  you would enable HA you wouldn't be able to do so with xhad, thus xhad is not a valid choice anymore.